fuel-asm/src/macros.rs

```diff
@@ -1,42 +1,678 @@
+        /// Access the ID for register D.
+        pub fn rd(&self) -> RegId {
+            unpack::rd_from_bytes(self.0)
+        }
+    };
+    (impl_op_accessors [RegId RegId Imm12]) => {
+        impl_instructions!(impl_op_accessors [RegId RegId]);
+        /// Access the 12-bit immediate value.
+        pub fn imm12(&self) -> Imm12 {
+            unpack::imm12_from_bytes(self.0)
+        }
+    };
+    (impl_op_accessors [RegId Imm18]) => {
+        impl_instructions!(impl_op_accessors [RegId]);
+        /// Access the 18-bit immediate value.
+        pub fn imm18(&self) -> Imm18 {
+            unpack::imm18_from_bytes(self.0)
+        }
+    };
+    (impl_op_accessors [Imm24]) => {
+        /// Access the 24-bit immediate value.
+        pub fn imm24(&self) -> Imm24 {
+            unpack::imm24_from_bytes(self.0)
+        }
+    };
+    (impl_op_accessors []) => {};
+
+    // Generate a method for converting the instruction into its parts.
+    (impl_op_unpack [RegId]) => {
+        /// Convert the instruction into its parts.
+        pub fn unpack(self) -> RegId {
+            unpack::ra_from_bytes(self.0)
+        }
+    };
+    (impl_op_unpack [RegId RegId]) => {
+        /// Convert the instruction into its parts.
+        pub fn unpack(self) -> (RegId, RegId) {
+            unpack::ra_rb_from_bytes(self.0)
+        }
+    };
+    (impl_op_unpack [RegId RegId RegId]) => {
+        /// Convert the instruction into its parts.
+        pub fn unpack(self) -> (RegId, RegId, RegId) {
+            unpack::ra_rb_rc_from_bytes(self.0)
+        }
+    };
+    (impl_op_unpack [RegId RegId RegId RegId]) => {
+        /// Convert the instruction into its parts.
+        pub fn unpack(self) -> (RegId, RegId, RegId, RegId) {
+            unpack::ra_rb_rc_rd_from_bytes(self.0)
+        }
+    };
+    (impl_op_unpack [RegId RegId Imm12]) => {
+        /// Convert the instruction into its parts.
+        pub fn unpack(self) -> (RegId, RegId, Imm12) {
+            unpack::ra_rb_imm12_from_bytes(self.0)
+        }
+    };
+    (impl_op_unpack [RegId Imm18]) => {
+        /// Convert the instruction into its parts.
+        pub fn unpack(self) -> (RegId, Imm18) {
+            unpack::ra_imm18_from_bytes(self.0)
+        }
+    };
+    (impl_op_unpack [Imm24]) => {
+        /// Convert the instruction into its parts.
+        pub fn unpack(self) -> Imm24 {
+            unpack::imm24_from_bytes(self.0)
+        }
+    };
+    (impl_op_unpack []) => {};
+
+    // Generate a shorthand free function named after the $op for constructing an `Instruction`.
+    (impl_op_constructor $doc:literal $Op:ident $op:ident [RegId]) => {
+        #[doc = $doc]
+        pub fn $op(ra: u8) -> Instruction {
+            $Op::new(check_reg_id(ra)).into()
+        }
+    };
+    (impl_op_constructor $doc:literal $Op:ident $op:ident [RegId RegId]) => {
+        #[doc = $doc]
+        pub fn $op(ra: u8, rb: u8) -> Instruction {
```

Comment:
I think we should be taking the typed `RegId` here instead of `u8`

Reply:
The idea behind this constructor was to be a more "ergonomic" alternative, solely for handwriting ASM (i.e. writing tests - see [this section of the doc comment](https://github.com/FuelLabs/fuel-vm/blob/cf52dc7eedd2e7f55ab5467e92e33bc27396cd00/fuel-asm/src/macros.rs#L168-L190)). You can still use the `$Op::new` constructor directly which checks `RegId` at compile-time rather than runtime (e.g. as the compiler will do), but it is a little verbose for tests.

I'm open to making this change if you still think it's worth taking `RegId` here, but it will mean updating loads of tests :sweat_smile:

Reply:
Right. The idea of updating this came to me from doing lots of test changes, which require changing lines like `Opcode::RET(REG_ONE)` to `op::ret(REG_ONE.into())` or `op::ret(1)`. The latter seems quite misleading; changing it to `op::ret(2)` would return the value of `$of` instead of `2`.  The nice thing with typed registers is that we could instead have `op::ret(RegId::ONE)` instead.

Reply:
@Dentosal ahh I can see your point. I have no major preference either way, will leave the call up to you!

Reply:
After wasting a couple of hours on trying to figure how to update the tests without spending days on rewriting them, I've given up on this. The current way of things is good enough.

Reply:
I made the shorthands take `CheckRegId` trait instead of `u8` for `RegId` and impl'd that for both `u8` and `RegId`, so that at least we don't have to do `REG_ZERO.into()` at every callsite, but `RegId` can be passed directly.